app/src/server/api/external/v1Api.router.ts

```diff
@@ -115,62 +103,89 @@ export const v1ApiRouter = createOpenApiRouter({
+          code: "NOT_FOUND",
         });
       }
 
-      try {
-        const completion = await getCompletion2(typedFT, inputPayload);
-        const inputTokens = completion.usage?.prompt_tokens ?? 0;
-        const outputTokens = completion.usage?.completion_tokens ?? 0;
+      let completion: ChatCompletion | Stream<ChatCompletionChunk>;
 
-        const cost = calculateCost(typedFT, 0, inputTokens, outputTokens);
+      if (fineTune) {
+        const typedFT = typedFineTune(fineTune);
 
-        // Don't `await` this to minimize latency
-        prisma.usageLog
-          .create({
-            data: {
-              fineTuneId: fineTune.id,
-              type: UsageType.EXTERNAL,
-              inputTokens,
-              outputTokens,
-              cost,
-            },
-          })
-          .catch((error) => captureException(error));
+        if (typedFT && typedFT.projectId !== key.projectId) {
+          throw new TRPCError({
+            message: "The model does not belong to this project",
+            code: "FORBIDDEN",
+          });
+        }
 
-        posthogServerClient?.capture({
-          distinctId: fineTune.projectId,
-          event: "fine-tune-usage",
-          properties: {
-            model: inputPayload.model,
-            inputTokens,
-            outputTokens,
-            cost,
-          },
-        });
+        try {
+          completion = await getCompletion2(typedFT, inputPayload);
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get fine-tune completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+          });
+        }
+      } else {
+        try {
+          completion = await getOpenaiCompletion(key.projectId, {
+            ...inputPayload,
+            // @ts-expect-error gotta love overloads
+            stream: input.stream,
+          });
+        } catch (error: unknown) {
+          console.error(error);
+          throw new TRPCError({
+            message: `Failed to get OpenAI completion: ${(error as Error).message}`,
+            code: "BAD_REQUEST",
+            cause: error as Error,
+          });
+        }
+      }
 
-        return completion;
-      } catch (error: unknown) {
-        console.error(error);
-        throw new TRPCError({
-          message: `Failed to get completion: ${(error as Error).message}`,
-          code: "BAD_REQUEST",
+      const logRequest = ctx.headers["op-log-request"] === "true";
+      let tags: Record<string, string> = {};
+      if (ctx.headers["op-tags"]) {
+        try {
+          tags = JSON.parse(ctx.headers["op-tags"] as string);
+        } catch (error: unknown) {
+          // ignore
+        }
+      }
+
+      if (completion instanceof Stream) {
+        // split stream to avoid locking the read mutex
+        const [recordStream, outputStream] = completion.tee();
+        void captureRecordUsage({
```

Comment:
Shouldn't this go inside `recordUsage`?

Reply:
I don't see a problem with combining recordUsage and captureRecordUsage. Let's do it.